crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -87,78 +170,128 @@ impl MemoryPool {
-            usize::try_from(memory_size + tunables.static_memory_offset_guard_size)
-                .map_err(|_| anyhow!("memory reservation size exceeds addressable memory"))?;
-
-        assert!(
-            memory_and_guard_size % crate::page_size() == 0,
-            "memory size {} is not a multiple of system page size",
-            memory_and_guard_size
-        );
-
-        let max_total_memories = config.limits.total_memories as usize;
-        let initial_memory_offset = if tunables.guard_before_linear_memory {
-            usize::try_from(tunables.static_memory_offset_guard_size).unwrap()
-        } else {
-            0
+        // Create a slab layout and allocate it as completely inaccessible
+        // region to start--`PROT_NONE`.
+        let constraints = SlabConstraints {
+            max_memory_bytes: max_memory_bytes as usize,
+            num_memory_slots: config.limits.total_memories as usize,
+            num_pkeys_available: pkeys.len(),
+            guard_bytes: tunables.static_memory_offset_guard_size as usize,
+            guard_before_slots: tunables.guard_before_linear_memory,
         };
-
-        // The entire allocation here is the size of each memory (with guard
-        // regions) times the total number of memories in the pool.
-        //
-        // Note, though, that guard regions are required to be after each linear
-        // memory. If the `guard_before_linear_memory` setting is specified,
-        // then due to the contiguous layout of linear memories the guard pages
-        // after one memory are also guard pages preceding the next linear
-        // memory. This means that we only need to handle pre-guard-page sizes
-        // specially for the first linear memory, hence the
-        // `initial_memory_offset` variable here. If guards aren't specified
-        // before linear memories this is set to `0`, otherwise it's set to
-        // the same size as guard regions for other memories.
-        let allocation_size = memory_and_guard_size
-            .checked_mul(max_total_memories)
-            .and_then(|c| c.checked_add(initial_memory_offset))
-            .ok_or_else(|| {
-                anyhow!("total size of memory reservation exceeds addressable memory")
-            })?;
-
-        // Create a completely inaccessible region to start
-        let mapping = Mmap::accessible_reserved(0, allocation_size)
+        let layout = calculate(&constraints)?;
+        log::debug!("creating memory pool: {constraints:?} -> {layout:?}");
+        let mut mapping = Mmap::accessible_reserved(0, layout.total_slab_bytes)
             .context("failed to create memory pool mapping")?;
 
+        // Then, stripe the memory with the available protection keys. This is
+        // unnecessary if there is only one stripe color.
+        if layout.num_stripes >= 2 {
+            let mut cursor = layout.pre_slab_guard_bytes;
+            let pkeys = &pkeys[..layout.num_stripes];
```

Comment:
Haven't finished reading through stuff but: will `pkeys.len()` and `layout.num_stripes` ever be different? If so, I find that pretty surprising. What's the motivation for that?

Reply:
Yes: imagine the case where we (for some odd reason) only want a pooling allocator to have five memory slots. We can only have a maximum of five stripes. If we have fifteen keys available, we need to pick a subset of them for striping.